src/SSHDebugPS/VSOperationWaiter.cs

```diff
@@ -22,15 +22,21 @@ internal static class VSOperationWaiter
         /// <exception cref="OperationCanceledException">Wait was canceled and 'throwOnCancel' is true</exception>
         public static bool Wait(string actionName, bool throwOnCancel, Action action)
         {
-            // TODO: Implement cancellationToken once remoteSystem.Connect(connectionInfo) handles it.
-            // Currently, the TWD will show up without a cancelation button but it will end in 30 seconds.
-            ThreadHelper.JoinableTaskFactory.Run(
-                actionName,
-                async (progress) =>
-                {
-                    await Task.Run(action);
-                }
-            );
+            try
+            {
+                //TODO: Once 'action' is cancelable, provide token to the action instead.
+                ThreadHelper.JoinableTaskFactory.Run(
+                    actionName,
+                    async (progress, cancellationToken) =>
+                    {
```

Comment:
I don't think you want to use `Task.Run(...).Wait()` as this will block two different threads.

I think you want something like this:
```
        // Kicks off the specified action, and returns a Task which is completed when either the action finishes or the
        // cancellation token is set.
        static Task RunUntilCancelled(Action action, CancellationToken cancellationToken)
        {
            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();

            CancellationTokenRegistration cancellationTokenRegistration = cancellationToken.Register(() =>
            {
                // Mark the task returned from this method as cancelled, but we cannot cancel the underlying operation,
                // so we will just let it do nothing.
                tcs.TrySetCanceled();
            });

            Task runResult = Task.Run(action);
            runResult.ContinueWith((Task t) =>
            {
                tcs.TrySetResult(null);
                cancellationTokenRegistration.Dispose();
            });

            return tcs.Task;
        }
```

Reply:
Wow. That's really complicated. I agree you don't want to use `Task.Wait()` here though, since that's just what JTF.Run is supposed to take care of for you. So just replace your Wait() line with this:
```cs
await Task.Run(action).WithCancellation(cancellationToken);
```

That extension method will take care of the rest for you to get what you wanted. But as your code comment suggests, this still leaves `action` running on a background thread, with the risk perhaps that the user will repeat the operation and you'll end up stacking up multiple threads all running `action`. So you should change `action` to take a `CancellationToken` so that your cancellation is _real_ as soon as you can.

Reply:
Glad you are on the review so we didn't need to duplicate `WithCancellation` :)